### ISSUE TITLE
live stream not working, home overview updates fine

### ISSUE THREAD

**Issue by tmur11** [KhaosT/homebridge-camera-ffmpeg#12]
I'm having difficulties with the live stream not working. The overview will show me stills of the camera but the live stream will not load. Just sits spinning and spinning. I've confirmed i can play my stream URL with ffplay at 1280x960 @ 25fps. I'm running a mac VM with 4gb of memory.

Any input would be much appreciated.

**Comment by cmh716** [comment]
I have this same issue.  I don't even see the ffmpeg process launch on streaming.  I only see it launch to make the initial jpeg screenshot.  Looks like many of us have the same issue.

**Comment by KhaosT** [comment]
Try running the command directly with ffmpeg in Terminal I guess. When you tries to start the stream, the plugin will print out a string in Terminal which contains the parameters for ffmpeg. Try to run ``ffmpeg string_in_console`` and see what ffmpeg says about it.

**Comment by cmh716** [comment]
KaosT - See that's the issue for me.  Even when I start the stream, nothing prints out in the console.  It's almost like the streaming never gets initiated.

**Comment by KhaosT** [comment]
try to run ``DEBUG=* homebridge your_params`` and start stream again?

**Comment by cmh716** [comment]
Well, this is VERY interesting.  I upgraded my iPhone yesterday from an iPhone 6s Plus to an iPhone 7 Plus and suddenly it started working this morning.  I changed nothing else on the homebridge side.  Very strange.

**Comment by KhaosT** [comment]
lol that's strange... Let me know if you see the same problem in the future.

**Comment by cmh716** [comment]
Ok, I found that if I create a new Camera device in the json with a high resolution such as 1600x1200, streaming never gets initiated.  Creating a new device that's only 640x480 worked.  Just an idea that perhaps this is what others are experiencing.

**Comment by JangoBritt** [comment]
I had a similar problem. I stayed with 1280x720, but as my camera is only recording 15fps anyway, I reduced the MaxFPS in the config, and it's made a huge difference to both the home snapshots, and to the live stream

**Comment by tmur11** [comment]
@cmh716 that is very interesting that the upgrade from 6s Plus to 7 Plus improved you situation. I actually upgrade myself last night from iphone 6 to 7 Plus but it didn't improve my situation any.

I feel like my ffmpeg process isn't launching either as when i play the stream with ffplay i get significant CPU usage but running through homebridge no CPU usage increased is seen.

@KhaosT i will run 'ffmpeg string_in_console' tonight when I get a chance and post my results. I've tried it a few days ago and it did appear to be functioning/connecting to my stream.

@cmh716 you're saying you made a new camera device with a 640x480 resolution even though your camera stream was truly a 1280x720 stream?

@JangoBritt If i set my FPS to say 15 would they stream even work if my camera is configured for 25fps (I'd try this with ffplay if i knew how to pass it parameters such as resolution and fps)
  
Very impressive work here guys!

**Comment by JangoBritt** [comment]
@tmur11 I guess ffmpeg would read the input of 25 fps, and output at 15fps, yes... but it depends where your bottleneck is. If the problem is your VM not reading the stream, or the processing power on the VM, then it won't help... but if it's your stream passing through to homebridge, then it might do.

**Comment by tmur11** [comment]
@JangoBritt I can play my stream with `ffplay` buttery smooth but homebridge doesn't appear to be even trying. I tried reducing my resolution and FPS along with making the change at the camera level with no luck. 

@KhaosT i tried `ffmpeg my_stream_url` command and it appears to resolve my camera's stream size and fps successfully. just no luck with the live stream loading. Running `DEBUG=* homebridge` and `DEBUG=* homebridge -U ~/.homebridge-camera-only` (so that I can use a config file for a camera platform only). I don't see anything with DEBUG when attempting to load the live stream.

@cmh716 did you set your new iphone 7 plus up as a new iphone or restore from your old? reason i ask is i wonder if the new 2FA security is related to my issue. I've tried enabling icloud keychain (which from what i understand enables the Two Factor Authentication.

Are you guys running this plugin along side others?

**Comment by tmur11** [comment]
I was wrong, here's what i get when i ran `DEBUG=* homebridge`

  EventedHTTPServer [::ffff:10.172.2.88] HTTP request: /characteristics?id=1.14 +7s
  HAPServer [C8:57:48:7C:2A:0C] HAP Request: GET /characteristics?id=1.14 +1ms
  Accessory [Test Camera] Getting value for Characteristic "Streaming Status" +2ms
  Accessory [Test Camera] Got Characteristic "Streaming Status" value: AQEA +1ms
  EventedHTTPServer [::ffff:10.172.2.88] HTTP Response is finished +0ms
  EventedHTTPServer [::ffff:10.172.2.88] HTTP request: /characteristics +5ms
  HAPServer [C8:57:48:7C:2A:0C] HAP Request: PUT /characteristics +0ms
  Accessory [Test Camera] Processing characteristic set: [{"aid":1,"iid":14,"ev":true}] +1ms
  Accessory [Test Camera] Registering Characteristic "Streaming Status" for events +0ms
  EventedHTTPServer [::ffff:10.172.2.88] HTTP Response is finished +0ms
  EventedHTTPServer [::ffff:10.172.2.88] New connection from client +32ms
  EventedHTTPServer [::ffff:10.172.2.88] HTTP server listening on port 59167 +0ms
  EventedHTTPServer [::ffff:10.172.2.88] HTTP request: /pair-verify +18ms
  HAPServer [CE:22:3D:C1:CE:30] HAP Request: POST /pair-verify +0ms
  EventedHTTPServer [::ffff:10.172.2.88] HTTP Response is finished +1ms
  EventedHTTPServer [::ffff:10.172.2.88] Client connection closed +8ms
  EventedHTTPServer [::ffff:10.172.2.88] HTTP connection was closed +2ms
  EventedHTTPServer [::ffff:10.172.2.88] HTTP server was closed +1ms

**Comment by tmur11** [comment]
Okay some updates... I've managed to get the live streaming working by trying every resolution, fps, rtsp streams my camera had available. It does seem to be related resources available to the machine. I believe that ffmpeg process won't even start/crashes if the task is too CPU intensive. I have a 640*480 @ 15fps working but the performance is terrible when viewing the live stream, CPU usage ~90-120%

Hope that helps everyone else struggling with this issue and maybe we can work to improve this plugins performance.

**Comment by cmh716** [comment]
tmur11 - my iPhone was a restore but I'm think it's coincidental.  These are my findings:


If I create a device called "Driveway" that's 1600x1200 -- no stream at all.  nothing in the homebridge console indicating streaming has started.  If I then change the resolution to 640x480 -- exact same behavior.  If I then alter the name as well as the resolution and add the new device the my iPhone it works.  So "Driveway Cam" 640x480

So long story short, it looks like I'm unable to reduce the resolution on an existing device and instead need to create a NEW device to get it to work.  I think the new iPhone didn't have the remnants of the old deleted device causing trouble.

**Comment by gijoecool** [comment]
It seems that no matter how I have my cameras configured ffmpeg is started with the same scale. If I am on wifi, then scale=1280:720. If I am off wifi, then scale=640:480.

Is this something that iOS requests? Or maybe there is a bug here...

**Comment by tmur11** [comment]
@gijoecool I'm also seeing the same thing. with `DEBUG=* homebridge` i see a scale=1280:720 regardless  but CPU usage does appear to be affected depending on what resolution you have setup in the .config. As  @cmh716 has found a new device needs to be created before your .config files take affect. Shows scale=1280:720 regardless though.

**Comment by tmur11** [comment]
@KhaosT any thoughts on how we could achieve some performance improvements in our configurations? My live streams even turned way down are still almost usable. I have tried on a couple machines with more resources and get the same poor results.

**Comment by KhaosT** [comment]
I'm not really sure in this case. It's something highly dependent on the machine and the video source. I'd encourage you search online for optimizing ffmpeg speed or something...

**Comment by tmur11** [comment]
the thing is if I play the stream simply in `ffplay` the stream uses resources but frame rate is smooth at 25fps but homebridge with max frame rate even at 15fps and reduced resolution the live view is almost useless.

**Comment by KhaosT** [comment]
Then i think maybe the conversion part is taking too much CPU time? What's the CPU usage during the live streaming?

**Comment by benklinger** [comment]
I have the same problem. Updated the FPS to 1 and it started to work, but misses alot and gets stuck.
No problem when viewing it in VLC.

**Comment by EvianZhow** [comment]
@KhaosT Sorry to disturb you again. Camera snapshot works fine, I've tried to add `console.log();` to snapshot functions, and it turned out to be fine. Printed as expected. But live stream is not working and no ffmpeg parameters showing from the logs. Maybe there is a problem related with iOS?

**Comment by KhaosT** [comment]
@EvianZhow You need to grab logs from iOS device to figure out why HomeKit doesn't request stream from accessory. On macOS Sierra, you can connect your iOS device to your Mac, open Console.app and select the iOS device on the sidebar. Then try to start streaming and see what homed complain about the stream.

**Comment by EvianZhow** [comment]
@KhaosT Sure, I've post some parts of the logs under https://github.com/KhaosT/homebridge-camera-ffmpeg/issues/30, last comment. I think I should investigate once more. Thanks and keep updated!
